### PR TITLE
test: LbaasV2 method buildPoolCreateOpt

### LINKED ISSUE #2400: [tests] Add unit tests for the loadbalancer.go functions
/kind feature

**What happened**:

The cloud-provider-openstack code most related to this project lives in [loadbalancer.go](https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/openstack/loadbalancer.go) which is lacking almost all unit tests. This task is to write tests for all of the functions there. Please note that it is untrivial to create tests for methods using gophercloud library, so calling the OpenStack APIs.

**List of Tests**:
- [x] getSecurityGroupName
- [x] getListenerProtocol
- [x] GetLoadBalancerName
- [x] nodeAddressForLB
- [x] getStringFromServiceAnnotation
- [x] getIntFromServiceAnnotation
- [x] getBoolFromServiceAnnotation
- [x] buildMonitorCreateOpts
- [x] buildPoolCreateOpt
- [x] buildBatchUpdateMemberOpts
- [x] buildListenerCreateOpt
- [x] getMemberSubnetID
- [x] getSubnetID
- [x] getNetworkID
- [x] checkListenerPorts
- [x] updateServiceAnnotations
- [x] createLoadBalancerStatus
- [ ] GetLoadBalancerSourceRanges

### PR DESCRIPTION
**Pull request by majorchork**
**What this PR does / why we need it**:
this pr adds unit test for buildPoolCreateOpt LbaasV2 method in loadbalancer.go 
**Which issue this PR fixes(if applicable)**:
refers #2400 

**Special notes for reviewers**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by sakshi-1505 — COMMENTED** [review]
Can you remove `fixes` from the PR description or else this PR will close the outstanding issue?

File: pkg/openstack/loadbalancer_test.go [hunk 1]
```diff
@@ -534,3 +535,195 @@ func Test_getListenerProtocol(t *testing.T) {
...
+				Protocol:    pools.ProtocolHTTP,
+				LBMethod:    "ROUND_ROBIN",
+				Persistence: &pools.SessionPersistence{Type: "SOURCE_IP"},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := tt.args.lbaasV2.buildPoolCreateOpt(string(tt.args.protocol), tt.args.service, tt.args.svcConf); !assert.Equal(t, got, tt.want) {
```
Comment (sakshi-1505) [inline pkg/openstack/loadbalancer_test.go:724]: How about we use reflect.DeepEqual for matching struct & custom data objects @dulek ?

**Comment by majorchork** [comment]
> Can you remove `fixes` from the PR description or else this PR will close the outstanding issue?

Once a PR is merged the issue should be closed, and a new one created as all the marked tests need to be reopened

File: pkg/openstack/loadbalancer_test.go (hunk 1 above)
Comment (majorchork) [reply pkg/openstack/loadbalancer_test.go]: why is this here @sakshi-1505 ?

**Comment by sakshi-1505** [comment]
> > Can you remove `fixes` from the PR description or else this PR will close the outstanding issue?
> 
> Once a PR is merged the issue should be closed, and a new one created as all the marked tests need to be reopened

I didn't understand this but we don't want that parent tracking issue to get closed which will be if this PR gets merged. Issue #2400 not only tracks this tests but a lot of other unit tests where different contributors are involved which is why I suggested to use `refer` instead of close.

File: pkg/openstack/loadbalancer_test.go [hunk 2]
```diff
@@ -593,3 +594,195 @@ func Test_getIntFromServiceAnnotation(t *testing.T) {
...
+				LBMethod:    "ROUND_ROBIN",
+				Persistence: &pools.SessionPersistence{Type: "SOURCE_IP"},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := tt.args.lbaasV2.buildPoolCreateOpt(tt.args.protocol, tt.args.service, tt.args.svcConf)
+			if !assert.Equal(t, got, tt.want) {
```
Comment (dulek) [inline pkg/openstack/loadbalancer_test.go:783]: You don't need to check result of `assertt.Equal()`, it'll report the issue on its own.

File: pkg/openstack/loadbalancer_test.go [hunk 3]
```diff
@@ -593,3 +594,195 @@ func Test_getIntFromServiceAnnotation(t *testing.T) {
...
+					healthMonitorMaxRetries:     1,
+					healthMonitorMaxRetriesDown: 1,
+					healthCheckNodePort:         2,
+					enableProxyProtocol:         true,
+				},
+				lbaasV2: &LbaasV2{
+					LoadBalancer{
+						opts: LoadBalancerOpts{
+							LBProvider: "ovn",
+							LBMethod:   "ROUND_ROBIN",
```
Comment (dulek) [inline pkg/openstack/loadbalancer_test.go:628]: Not really important, feel free to ignore, but this combinantion wouldn't work in real world, `ovn` only works with `SOURCE_IP_PORT`.

File: pkg/openstack/loadbalancer_test.go [hunk 4]
```diff
@@ -593,3 +594,195 @@ func Test_getIntFromServiceAnnotation(t *testing.T) {
...
+			args: args{
+				protocol: "TCP",
+				svcConf: &serviceConfig{
+					keepClientIP:                true,
+					tlsContainerRef:             "tls-container-ref",
+					healthMonitorDelay:          1,
+					healthMonitorTimeout:        1,
+					healthMonitorMaxRetries:     1,
+					healthMonitorMaxRetriesDown: 1,
+					healthCheckNodePort:         2,
```
Comment (dulek) [inline pkg/openstack/loadbalancer_test.go:621]: I think you don't need to specify these, the function doesn't seem to use them? Same goes for other cases.

**Review by dulek — CHANGES_REQUESTED** [review]

**Comment by dulek** [comment]
/retest

Error's unrelated.

File: pkg/openstack/loadbalancer_test.go (hunk 3 above)
Comment (majorchork) [reply pkg/openstack/loadbalancer_test.go:628]: noted, but. i'll update it just so that it's not misleading to people reading it in the future

**Comment by majorchork** [comment]
@dulek all conflicts have been resolved

**Comment by majorchork** [comment]
Hello @jichenjc there was a small update to the function by @kayrus, so I had to update the tests, please re-add lgtm

**Comment by jichenjc** [comment]
/lgtm

:)

**Pull request merged**